gammapy/detect/cwt.py

```diff
@@ -57,200 +62,552 @@ def difference_of_gauss_kernel(radius, scale_step, n_sigmas=8):
+                        log.info('Remove significant pixels island {0} from support'.format(struct_label + 1))
+                        mask[coords] = False
 
-    - ``total_background = self.model + self.background + self.approx``
+            log.info('Update support for scale {:.2f}'.format(self.kernels.scales[idx_scale]))
+            data.support[idx_scale] |= mask
 
-    The ``compute_support_peak`` step does the following:
+    def _inverse_transform(self, data):
+        """
+        Do the inverse transform (reconstruct the image).
+
+        TODO: describe better what this does.
+
+        Parameters
+        ----------
+        data : `~gammapy.detect.CWTData`
+            Images for inverse transform.
+        """
+        data.mysterious_result = np.sum(data.support * data.transform, axis=0)
+        log.info('Update model')
+        data.model += data.mysterious_result * (data.mysterious_result > 0)
 
-    - computes significance ``sig = transform / error``
-    - support : 3D cube exclusion mask
-      - filled as ``sig > nsigma``
+    def _is_converged(self, data):
+        """
+        Check if the algorithm has converged on current iteration.
 
-    The ``inverse_transform`` step does the following:
+        TODO: document metric used, but not super important.
 
-    - model : 2D image, init to 0
-        - Fill ``res = np.sum(self.support * self.transform, axis=0)``
-        - Fill ``self.model += res * (res > 0)``
-        - What is this??
+        Parameters
+        ----------
+        data : `~gammapy.detect.CWTData`
+            Images after iteration.
 
+        Returns
+        -------
+        answer : boolean
+            Answer if CWT has converged.
+        """
+        answer = False
+        residual = data.counts - (data.model + data.approx)
+        variance = residual.var()
+        if self.previous_variance is not None:
+            variance_ratio = abs((self.previous_variance - variance) / self.previous_variance)
+            answer = variance_ratio < self.tol
+            log.info('Variance ratio: {:.7f}'.format(variance_ratio))
+        self.previous_variance = variance
+        log.info('')
+        return answer
+
+    def analyze(self, data):
+        """
+        Run iterative filter peak algorithm. The algorithm modifies the original data.
 
-    - max_scale_image : 2D image, value = index of scale where emission is most significant
+        Parameters
+        ----------
+        data : `~gammapy.detect.CWTData`
+            Input images.
+        """
+        if self.history is not None:
+            self.history = [copy.deepcopy(data)]
+
+        for n_iter in range(self.max_iter):
+            log.info('Start iteration {0}'.format(n_iter + 1))
+            self._execute_iteration(data=data)
+            if self.history is not None:
+                self.history.append(copy.deepcopy(data))
+            if self._is_converged(data=data):
```

Comment:
I think the class would be simpler if you don't make `previous_variance` an attribute on `self`, but instead a local variable here in `analyze`.

All this method does is book-keeping to run iteratively, so I think doing the stopping criterion here is a good fit.

The utility function you call should then just compute the number you use as stopping criterion.
Or, if you really think it makes sense, pass the `previous_variance` in from here, but don't store it as a CWT attribute.